### ISSUE TITLE
Technical Query: How to associate an icosahedron level with an equirectangular image size?

### ISSUE THREAD

**Issue by snitch30** [meder411/Tangent-Images#6]
Hi Marc, 

I have been reading your article [Medium- 360 Computer Vision Problems Article](https://medium.com/@meder411/360-computer-vision-problems-solutions-and-a-suggestion-for-the-path-ahead-438831c6277b). Really an interesting article. I was also able to generate tangent images without issues.

However, in the article (as seen in the image below), you state that "a 4K equirectangular image is equivalent to a level 10 icosahedron"? Is there any mathematical equivalent or relation to calculate the same for other resolutions.

> Screenshot Begin

<img width="725" alt="Screenshot 2020-08-28 at 11 05 12 PM" src="https://user-images.githubusercontent.com/24635047/91598256-f396fb80-e982-11ea-86f2-9267c2a3f37b.png">

> Screenshot End

I understand how 2^(10-1) results in 80 512x512 image. I am not sure about how you arrived at '10' (I might be missing something). Thanks in advance.

**Comment by meder411** [comment]
That relationship arises from the pixel-to-vertex or pixel-to-face analogy. Unfortunately, because the icosahedron has fixed number of elements at each level based on the subdivision algorithm, there is no direct relation between the number of vertices/faces of an icosahedron and the number of pixels in an image. Because of the factor of four relationship in subdivision and down/upsampling that we still want to preserve in the CNN paradigm, it is not exact. But if we assume a continuous image space (using some interpolant for areas between pixel centers), we can still get by with minimal issues.

How we arrive at this relation is by choosing the a subdivision level with the number of faces (20*4^s) that is closest to the number of pixels in the image, without going under. Effectively, we end up having to oversample the image a little bit.

For example, a level 10 icosahedron has 20*4^10 = 20,971,520 faces.
A 4000x2000 equirectangular image has 8,000,000 pixels.
So it seems quite a bit larger. But if we were to use a level 9 instead, we'd have 20*4^9 = 5,242,880 faces, leaving us with an undersampling.

This tracks with other sampling levels as well (keeping that 4x change in pixels per corresponding image).

I hope this clarifies things a bit

**Comment by snitch30** [comment]
Thanks for your response.

Yeah, that makes sense. After posting the question, when I kept searching for a relation, I assumed a similar relation with vertices was the reason for the approximation (that 4K ~ level 10).
For eg: level 8 icosahedron has 655,362 vertices and level 7 has 163,842 vertices.
So, for an image of resolution 512x1024 = 524,288 pixels, it was level 8 > image > level 7, so we take level 8 (closest).

However, I wasn't able to visualise the connection between vertices/faces and pixels (as I kept thinking there must exist one explicitly/direct). 
But now that you have pointed out,
> ... there is no direct relation between the number of vertices/faces of an icosahedron and the number of pixels in an image.

It kinda makes sense that the relationship is implicit (indirect).

Thanks.

**Comment by meder411** [comment]
Happy to help!

With tangent images, use the faces. The idea is that once we set a base level, we don't have to subdivide further, but it works out that the number of pixels in our set of tangent images is actually designed to be equivalent to the number of faces if we were to continue to subdivide.

For example, is s=2 and b=1, we would have our tangent image dimension, d=2^(2-1)=2, and thus our tangent images would be 2x2. Each tangent image is associated with a face of the base level 1 icosahedron. This means for each face, we have 4 pixels in our tangent image. Well, if we had just subdivided again (i.e. represented the s=2 resolution image with a level 2 icosahedron), each face would become 4 faces.

Basically we have this identity:

20*\(4^b) \* d \* d = 20\*(4^s)
20*\(4^b) \* 2^(s-b) \* 2(s-b) = 20\*(4^s)

Does this clear things up as to why to use the number of faces?
